src/storage/txn/commands/mod.rs

```diff
@@ -432,3 +562,29 @@ impl Debug for Command {
         self.command_ext().fmt(f)
     }
 }
+
+pub trait ReadCommand<S: Snapshot>: CommandExt {
+    fn process_read(self, snapshot: S, statistics: &mut Statistics) -> Result<ProcessResult>;
+}
+
+pub struct StorageToWrite<'a, S: Snapshot, L: LockManager, P: PdClient + 'static> {
+    snapshot: S,
+    lock_mgr: &'a L,
+    pd_client: Arc<P>,
+}
+
+pub struct WritingContext<'a> {
+    extra_op: ExtraOp,
+    statistics: &'a mut Statistics,
+    pipelined_pessimistic_lock: bool,
+}
```

Comment:
@nrc I'm not very sure about the names `StorageToWrite` and `WritingContext` here. Do you have any suggestiongs?

Reply:
How about `WriteContext` for the second and just passing the first as individual args? (Sorry for going back and forth, now that the struct is written out, it doesn't seem much like a coherent abstraction).